src/pallene/pallenelib.lua

```diff
@@ -90,12 +127,209 @@ static lua_Number  pallene_math_log(lua_Integer x, lua_Integer base);
+    pt_frame_t *tail = (pt_frame_t *) lua_topointer(L, -1);
+    lua_pop(L, 1);
+
+    if(tail != NULL)
+        tail -> line = line;
+}
+
+static void pallene_tracer_setline(pt_frame_t * restrict frame, int line) {
+    frame -> line = line;
+}
+
+static void pallene_tracer_frameexit(lua_State *L) {
+    pt_frame_t *head, *tail;
+
+    lua_getglobal(L, "__pallene_tracer_stack_head");
+    head = (pt_frame_t *) lua_topointer(L, -1);
+    lua_getglobal(L, "__pallene_tracer_stack_tail");
+    tail = (pt_frame_t *) lua_topointer(L, -1);
+    lua_pop(L, 2);
+
+    /* We are popping the very last frame. */
+    if(tail -> prev == NULL) {
+        tail = NULL;
+        head = NULL;
+
+        goto out;
+    }
+
+    tail -> prev -> next = NULL;
+    tail = tail -> prev;
+
+out:
+    /* Now update the registry. */
+    lua_pushlightuserdata(L, head);
+    lua_setglobal(L, "__pallene_tracer_stack_head");
+    lua_pushlightuserdata(L, tail);
+    lua_setglobal(L, "__pallene_tracer_stack_tail");
+}
+
+static int pallene_tracer_debug_traceback(lua_State *L) {
+    /* The debug traceback function frame. */
+    pt_frame_t self = {
+        .fn_name  = "pallene_tracer_debug_traceback",
+        .mod_name = "pallene_debug",
+        .line     = 0
+    };
+    pallene_tracer_frameenter(L, &self, (ptrdiff_t) pallene_tracer_debug_traceback);
+
+    const char *message = lua_tostring(L, 1);
+    fprintf(stderr, "Runtime error: %s\nStack traceback: \n", message);
+
+    lua_getglobal(L, "__pallene_tracer_stack_tail");
+    pt_frame_t *frame = (pt_frame_t *) lua_topointer(L, -1);
+    lua_pop(L, 1);
+
+    /* For context switch from Pallene to Lua or vice versa. */
+    /* We use the respective call-stack depending on the context.
+     * In Lua context we use the Lua call-stack and in Pallene context
+     * we use our self-maintained call stack. */
+    /* 1: Lua, 0: Pallene */
+    int context = 1;
+
+    /* Current level of depth we are at in the Lua call stack. */
+    int level   = 0;
+
+    /* Are we done iterating through all the call-frames in the Lua stack? */
+    bool gstack  = 0;
+
+    /* Which context we were in previously? */
+    /* The numeric representation is same as `context`. */
+    int prev_context = 1;
+
+    /* The frame signature. */
+    ptrdiff_t frame_sig = frame -> frame_sig;
+
+    while(true) {
+        /* If we are done iterating through the Lua stack and
+           there is no more frame left in Pallene Tracer stack,
+           we are done with the traceback. */
+        if(!gstack && frame == NULL)
+            break;
+
+        if(context) {
+            /* Get lua call stack information. */
+            lua_Debug ar;
+
+            if(!(gstack = lua_getstack(L, level, &ar)))
+                continue;
+
+            level++;
+
+            /* We need more info for a good traceback entry. */
+            lua_getinfo(L, "Slnt", &ar);
+
+            /* We have got a C frame. Time to make a context switch. */
```

Comment:
As pointed out by Lourival,  `lua_iscfunction` might be more robust than looking at the ar.what string.